core/src/main/scala/org/apache/spark/util/CompletionIterator.scala

```diff
@@ -25,11 +25,14 @@ private[spark]
 abstract class CompletionIterator[ +A, +I <: Iterator[A]](sub: I) extends Iterator[A] {
 
   private[this] var completed = false
-  def next(): A = sub.next()
+  private[this] var iter = sub
+  def next(): A = iter.next()
   def hasNext: Boolean = {
-    val r = sub.hasNext
+    val r = iter.hasNext
     if (!r && !completed) {
       completed = true
+      // reassign to release resources of highly resource consuming iterators early
+      iter = Iterator.empty.asInstanceOf[I]
```

Comment:
@szhem 
hi,
In your commet " reassign the reference of its sub-iterator to the empty iterator ..."

I don't understand why reassigning iter would take effect on original sub-iterator .
can you explain it more?

Reply:
@ChenjunZou, you can find the details in this [message](https://github.com/apache/spark/pull/23083#discussion_r236057101)

Reply:
Thanks, szhem :)
your UT explains all. 
at first I misunderstand sub as CompletionIterator(val sub)
BTW, sub is absolutely not type of strong reference, do you konw what is sub's reference type.